Allow synchronous router migration

rhafer opened the pull request:
The new --wait-for-ports causes neutron-ha-tool to wait for all
not-distributed router-ports to be back in the ACTIVE state on the
target agent when migrating a router. This can be used to avoid long
work queues to pile up in the neutron agents when migrating many routers
in one go. And by that reduce the downtime per migrated router.

rhafer committed:
neutron-ha-tool: Return full agent dicts instead of just the ids

Reworks the functions (and all their callers) for listing dead and alive
agents to return the full agent dict instead of just the ids. This is in
preparation for making the router migration sychronous. For that we will
need access to more agent attributes than just the id.

rhafer committed:
neutron-ha-tool: make router migration synchronous by default

The new --wait-for-ports option (enabled by default) causes
neutron-ha-tool to wait for all not-distributed router-ports to be back
in the ACTIVE state on the target agent when migrating a router. This
can be used to avoid long work queues to pile up in the neutron agents
when migrating many routers in one go. And by that reduce the downtime
per migrated router. This behaviour can be turned off with by using
the --no-wait-for-ports option.

Fixes: https://bugzilla.suse.com/show_bug.cgi?id=1012968

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -81,6 +81,9 @@ def parse_args():
                     help='Replicate DHCP configuration to all agents')
     ap.add_argument('--now', action='store_true', default=False,
                     help='Migrate Routers immediately without a delay.')
+    ap.add_argument('--wait-for-ports', action='store_true', default=False,
```
Better to default to `True`?

rhafer replied:
Hm, I not really sure yet what's the best default here. So I kept backwards compatible. 

Does it really make sense in the HA fail-over use case (when a single node, that is most likely offline already, needs to be evacuated quickly) to add all the waiting? I'd think that in that case it's preferable to have this script finish as fast as possible.  And after all this is still the default use-case of neutron-ha-tool.

aspiers replied:
It's more important that the script has the desired effect than that it finishes ASAP.  And we already know that in production clouds, allowing a backlog to build in the OVS agents and/or neutron servers can cause the whole thing to fall over in a spectacular heap.  Correctness should trump performance.

rhafer replied:
@aspiers I am still torn between the two possibilities here. You almost convinced me :).

I am just wondering if, by switching the default to ```True``` we make the release of this as an update to Cloud 6 more complicated than needed. AFAICS, if we don't switch the default we can just release this as is as an update to Cloud 6 (it shouldn't cause any difference in the behaviour, but provide what we need for the upgrade to work). And if we want to enable it for HA we'd just add an update to the ha barclamp that enables it  (maybe even providing a proposal attribute to switch) and reconfigures the pacemaker resource accordingly.

OTOH if we switch the default it seems to me (maybe I am wrong here) we would need to ship an update to the barclamp together with it. AFAIU when this feature is enable we need some config changes (timeouts at least) on the pacemaker resource.

So wouldn't we make our lives a bit easier by keeping the default?

aspiers replied:
> OTOH if we switch the default it seems to me (maybe I am wrong here) we would need to ship an update to the barclamp together with it. AFAIU when this feature is enable we need some config changes (timeouts at least) on the pacemaker resource.

@rhafer The `op start timeout` is already way too low as per [bsc#1011655 (neutron-ha-tool 120s start timeout too low for large clouds)](https://bugzilla.suse.com/show_bug.cgi?id=1011655), so we need an update to `crowbar-ha` anyway.

rhafer replied:
@aspiers Ok. I changed the default. With that I believe I addressed all the concerns you raised.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -219,18 +222,18 @@ def run(args):
         LOG.info("Performing L3 Agent Migration for Offline L3 Agents")
         errors = retry_with_backoff(l3_agent_migrate, args)(
-            qclient, args.noop, args.now)
+            qclient, args.noop, args.now, args.wait_for_ports)
```
Might be better to just pass `args` for simplicity, otherwise there's a lot of argument passing.  Not 100% sure though.

rhafer replied:
Yeah, I will most likely change that. Lastest for the other feature we need to add (ssh support to remove the namespaces).

aplanas replied:
IMHO is better to be explicit in the arguments passed, as is now.

rhafer replied:
Up to a certain number of arguments you're right I guess. But with the next change we will probably hit a limit.

aspiers replied:
Right.

rhafer replied:
@aspiers Are you ok if I address this with https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/8 ?

aspiers replied:
@rhafer Sure.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
+    """
+
+    # FIXME: Don't check the "network:router_interface_distributed" ports
+    # as those are not hosted on the network node
```
So this is currently broken for DVR?  That would probably be a blocker for merging :-/

rhafer replied:
I still need to look into that. Hence the "wip" label.

aspiers replied:
OK thanks.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
+
+    # FIXME: Don't check the "network:router_interface_distributed" ports
+    # as those are not hosted on the network node
+    router_port_list = qclient.list_ports(device_id=router_id, fields='id')
```
Out of curiosity how did you figure out this API?  The only doc I could find was http://docs.openstack.org/user-guide/sdk-neutron-apis.html which doesn't detail the parameters.

rhafer replied:
http://developer.openstack.org/api-ref/networking/v2/?expanded=list-ports-detail
https://wiki.openstack.org/wiki/Neutron/APIv2-specification#Filtering_and_Column_Selection

and the neutronclient source code :/

And yes the SDK API doc were not really helpful.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
+    port_ids = [r['id'] for r in router_port_list['ports']]
+
+    LOG.info("Wait for the ports of router_id=%s to be ACTIVE" % router_id)
+    while (maxtries > 0) and (len(port_ids) > 0):
```
Isn't this a bit overkill on the parentheses? :-)  An ignorable nitpick though.

rhafer replied:
Fixed

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
+    while (maxtries > 0) and (len(port_ids) > 0):
+        remaining = []
+        for port_id in port_ids:
+            port = qclient.show_port(port_id)['port']
```
Is it possible to retrieve this in a single API call, rather than one per port?  If so, that could reduce overall time taken for migrating all routers.

rhafer replied:
Hm, it's possible to get all router ports (and their details) in a single API call. Not sure which way is more efficient. But I guess you're right, getting them all at once will save us a couple of round trips to the server. I'll take a look at that.

rhafer replied:
I changed it to use list_port instead of fetching the single ports one by one. It does not have any measurable effect on migration time. But I guess just keep it that way now .

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
+        port_ids = remaining
+
+    if len(port_ids):
+        raise RuntimeError("Some ports are not ACTIVE on router_id=%s, ports=[%s]" %
```
I don't think you need to repeat the word `ports` here:

    raise RuntimeError("Some ports are not ACTIVE on router_id=%s: [%s]" %

Small nitpick but it's a pain when log lines are long enough to scroll off the right-hand edge of the terminal ...

rhafer replied:
Ack

rhafer replied:
Fixed

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
+            LOG.debug("Router port=%s on hostid=%s in status=%s" %
+                      (port['id'], port['binding:host_id'],
+                       port['status']))
+            if ((port['binding:host_id'] == target_host) and
```
Same nitpick about parentheses here; `and` is lower precedence than `==`.

rhafer replied:
Fixed

aspiers reviewed (changes requested):
Basically looks great, just a few nitpicks.  Please tell me next time you start working on the thing I said I was hoping to work on though ;-) Luckily on this occasion there wasn't much duplicated effort :-)

rhafer commented:
> Basically looks great, just a few nitpicks. Please tell me next time you start working on the thing I said I was hoping to work on though ;-) Luckily on this occasion there wasn't much duplicated effort :-)

Actually I told you. Though I must admit I didn't wait too long for an answer. Maybe you just missed it.

Itxaka commented on files/default/neutron-ha-tool.py:
```diff
@@ -339,34 +346,34 @@ def l3_agent_check(qclient):
             migration_count += 1
             LOG.warn("Would like to migrate router=%s to agent=%s",
-                     router_id, target_id)
+                     router_id, target['id'])
```
If `target` is `None` as set on line 367 this should raise a TypeError.

```
>>> import logging
>>> target = None
>>>logging.warn("Would like to migrate router=%s to agent=%s" % ("test", target['id']))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object has no attribute '__getitem__'
```

rhafer replied:
Should be fixed now.

Itxaka commented on files/default/neutron-ha-tool.py:
```diff
@@ -516,39 +523,39 @@ def replicate_dhcp(qclient, noop=False):
-        migrate_router(qclient, router_id, agent_id, target_id)
+        migrate_router(qclient, router_id, agent, target, wait_for_ports)
         return True
     except:
```
Not part of this PR, but this is too broad and we are the ones raising (some of) the exceptions so we should try to catch the 2 know raises that we emit and then fall back to catching everything.

matelakat replied:
FYI: Bare except is generally a bad thing to do https://docs.python.org/2/howto/doanddont.html#except

Itxaka commented on files/default/neutron-ha-tool.py:
```diff
@@ -559,31 +566,77 @@ def migrate_router(qclient, router_id, agent_id, target_id):
+    # FIXME: Don't check the "network:router_interface_distributed" ports
+    # as those are not hosted on the network node
+    router_port_list = qclient.list_ports(device_id=router_id, fields='id')
+    port_ids = [r['id'] for r in router_port_list['ports']]
```
consider using `port` instead of `r`, this code is read by humans ;)

rhafer replied:
It's gone now. I reworked the code a bit while addressing @aspiers feedback.

Itxaka reviewed (commented):
Only comments codewise, I have no idea if the code actually works :D

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -559,31 +566,81 @@ def migrate_router(qclient, router_id, agent_id, target_id):
+
+    active_port_ids = set()
+    remaining = ['dummy']
+    LOG.info("Wait for the ports of router_id=%s to be ACTIVE" % router_id)
```
nit: do not interpolate in logs: `LOG.info("Wait for the ports of router_id=%s to be ACTIVE", router_id)`

rhafer replied:
@aplanas Hm, since I am not that deep into python, could you share some more details here on why it's wrong?

Itxaka replied:
I think that one of the big issues is that if there is an encoding error, it wont raise an exception and break the code but instead it will be logged, as the interpolation is not done on the code but on the logger part, which can deal with that.

aplanas replied:
Not wrong at all, just not expected:

https://github.com/rhafer/cookbook-openstack-network/blob/529cb51e47f65060cc51dd936c2b8f6daff263ca/files/default/neutron-ha-tool.py#L228

You can read the documentation about this here:
https://docs.python.org/2/library/logging.html#logging.debug

rhafer replied:
Thanks. And fixed.

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -559,31 +566,81 @@ def migrate_router(qclient, router_id, agent_id, target_id):
+            else:
+                remaining.append(port['id'])
+
+        if len(remaining):
```
`if remaining`

rhafer replied:
Fixed

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -559,31 +566,81 @@ def migrate_router(qclient, router_id, agent_id, target_id):
+            time.sleep(1)
+            maxtries-=1
+
+    if len(remaining):
```
same

rhafer replied:
Fixed

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -585,8 +589,58 @@ def migrate_router(qclient, router_id, agent, target):
+    active_port_ids = set()
+    remaining = ['dummy']
+    LOG.info("Wait for the ports of router_id=%s to be ACTIVE" % router_id)
+    while (maxtries > 0 and len(remaining)):
```
`while remaining and maxtries:` can do the trick, also I was thinking how to populate `remaining` before, and move the `maxtries` check inside the while and ends it with a `break`

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -585,8 +589,58 @@ def migrate_router(qclient, router_id, agent, target):
+                         (port['id'], target_host))
+                active_port_ids.add(port['id'])
+            else:
+                remaining.append(port['id'])
```
IMHO this loop can be considered simplified with `set` operations. Basically we populate `remaining` (as a set) with `router_port_list['ports']` at the beginning, and we substract the set generated with the next `router_port_list['ports']` filtered with the condition on lines 628 and 629

rhafer replied:
```router_port_list['ports']``` is a list of dicts. You can't put dicts into a ```set``` (a dict is not hashable)

aplanas replied:
You are right, you need some data mangling to extract the new set that can be subtracted. I was referring to something like (not tested!!):

```python
remaining = {port['id'] for port in router_port_list['ports']}

while remaining:
    if not maxtries:
        break

    active_ports = {port['id'] for port in router_port_list['ports']
                    if port['binding:host_id'] == target_host and
                       port['status'] == 'ACTIVE')}
    remaining = remaining - active_ports
```

rhafer replied:
It's not that easy, actually. The list of ports needs to be refetched with every iteration, as the port state will change. And they might even change in number (e.g. users might have added/delete ports meanwhile).
Additionally (e.g. for logging/debugging purposes) I'd like to keep the state for which ports are already completed. While avoiding to log them over and over again. If I adjust your above suggestion to include those two things I think the resulting code has **not** much of an advantage over  the current code.

Edited: Added the missing "not".

aplanas replied:
OK for me

AbelNavarro replied:
i'm with @aplanas here. It would be better to fetch that list of ports just once and then keep the active/remaining for the loop. Notice that this is a wait for ports to be active after the migration and we were living without it for long. So if we skip a port that was recently added/modified by the user we shouldn't care that much.

rhafer replied:
@abel-navarro But we need to fetch the port list continuously otherwise, we will not get the state changes? The initial implementation I had pollled the ports one by one (using port_show), but that will cause a lot more round trips to the neutron server than just fetching all router ports. (See @aspiers comment and my replies on that: https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/7#discussion_r91616666)

AbelNavarro replied:
@rhafer I'm ok with the current implementation provide it has some significant advantage (maybe performance) over iterating on single ports. Otherwise I'd go to:

1. retrieve list of ports
2. iterate over ports that are not ACTIVE (polling single port)

rhafer replied:
@abel-navarro on pretty small setup (2 network nodes, 50 routers with 2 ports each, no floating IPs) I don't see any significant difference (perfomance-wise) between either approaches.
But I could image that in a setup with many ports to poll that the aproach with  polling singles ports is less performant (if just because of the many Rest-API roundtrips that are required for that)

aspiers replied:
The customer cloud in question has many ports on most of the routers I looked at.

rhafer replied:
Ok. I suggest to keep it to the current implementation for now.

aspiers commented:
Still think it should default to waiting, but willing to consider arguments to the contrary.

> Actually I told you. Though I must admit I didn't wait too long for an answer. Maybe you just missed it.

I was half-expecting you to say that, since as a rule it is much more likely that I missed something than that you communicated poorly ;-) In this case I now see a message in my IRC logs which I definitely missed at the time...

rhafer commented:
Hm, adding the WIP label back, because I still need to address the "FIXME" about DVR routers :/

aspiers reviewed (approved):
LGTM now - thanks!

aspiers commented:
Oh yeah, I forgot about the DVR routers FIXME ...

matelakat commented on files/default/neutron-ha-tool.py:
```diff
@@ -415,10 +422,10 @@ def l3_agent_migrate(qclient, noop=False, now=False):
         (migrations, errors) = \
             migrate_l3_routers_from_agent(
-                qclient, agent_id, agent_alive_list, noop)
+                qclient, agent, agent_alive_lqist, noop, wait_for_ports)
```
You have a typo here, that will break things `s/agent_alive_lqist/agent_alive_list`

rhafer replied:
@matelakat Should be fixed now. Thanks for noticing

matelakat reviewed (changes requested):
Please fix the typo

rhafer commented:
I've reworked the loop in ```wait_router_ports_active``` once more. I think the currently implementation is somewhat cleaner (and more a long the line to what @aplanas suggested). It also should handle distributed routers correctly now, by not looking at port that have ```binding:vif_type=="distributed"```

rhafer commented:
I noticed there is still one remaining bit. As floating IPs are not represented as ports on the router. We need to have a separate check for them to be migrated. I am currently working on that. Should have something ready soon.

rhafer committed:
neutron-ha-tool: Wait for floating IPs on router migration

This renames the --wait-for-ports option to --wait-for-router and
reworks it to also check for all floating IPs that are associated with a
migrated router to be back in the ACTIVE state again.

rhafer commented:
With the latest commit I think this change should be (feature) complete now. (So I removed the WIP label). Please review again. If preferred I can also squash the last three commit into one.

matelakat commented:
@rhafer looking at it - rebasing my branch on your work.

matelakat commented on files/default/neutron-ha-tool.py:
```diff
@@ -559,29 +574,88 @@ def migrate_router(qclient, router_id, agent_id, target_id):
+             "to be ACTIVE", router_id)
+    remaining_ports = ["dummy"]
+    remaining_fips = ["dummy"]
+    while maxtries:
```
this block will always add a one second sleep, no matter how fast my ports become active, as initially both `remaining_fips` and `remaining_ports` are true. If I understand it correctly, the intention is to first check the ports, and then the floating ips, right? The other problem, is that the minimal value for `maxtries` should be 2, because one is taken by checking the ports, the other one is taken by checking the ips.

Let's use `continue` if we see that `remaining_ports` has emptied!

Also, could we add a separate variable instead of mutating `maxtries`, it is quite weird to read `maxtries -= 1`, something like this would read better:

```python
remaining_iterations = maxtries
while remaining_iterations:
    ...
    remaining_iterations -= 1
```

rhafer replied:
> this block will always add a one second sleep, no matter how fast my ports become active, as initially both remaining_fips and remaining_ports are true.

Hm, true. Thanks for noticing that.

>  If I understand it correctly, the intention is to first check the ports, and then the floating ips, right? 

Yeah right, it doesn't make much sense to start checking on the floating IPs before the ports a back.

matelakat commented on files/default/neutron-ha-tool.py:
```diff
@@ -559,29 +574,88 @@ def migrate_router(qclient, router_id, agent_id, target_id):
+                         port['binding:host_id'] == target_host))
+            ]
+            LOG.debug("Ports not ACTIVE on router_id=%s: [%s]",
+                      router_id, ", ".join(remaining_ports))
```
add:
```python
if not remaining_ports:
    continue
```

matelakat commented on files/default/neutron-ha-tool.py:
```diff
@@ -559,29 +574,88 @@ def migrate_router(qclient, router_id, agent_id, target_id):
+                              if fip['status'] != 'ACTIVE']
+            LOG.debug("Floating IPs not active: [%s]",
+                      ", ".join(remaining_fips))
+
```
add:
```python
if not remaining_fips:
    break
```

matelakat commented on files/default/neutron-ha-tool.py:
```diff
@@ -559,29 +574,88 @@ def migrate_router(qclient, router_id, agent_id, target_id):
+            LOG.debug("Floating IPs not active: [%s]",
+                      ", ".join(remaining_fips))
+
+        if remaining_ports or remaining_fips:
```
We will no longer need this check I believe, we only need the sleep if the iteration is still non-zero.

matelakat reviewed (changes requested):
Could you please amend the belly of `wait_router_migrated` - so it doesn't do a delay every time?

rhafer committed:
neutron-ha-tool: Avoid unneeded sleep()s when waiting for router ports

This does a small optimization to wait_router_migrated() to skip the
sleep(1) call after all router_ports are up again and before starting
the checks on floating IPs.
Additionaly, to improve readability, this introduces the
remaining_iterations variable instead of directly manipulating maxtries.

rhafer commented:
> Could you please amend the belly of wait_router_migrated - so it doesn't do a delay every time?

@matelakat I believe I addressed all the issues you raised.

rhafer commented:
@aspiers are you still ok with this?

rhafer commented:
![](https://github.trello.services/images/mini-trello-icon.png) [bnc#1012968 [upgrade] router migration took too long during pre-upgrade of one node](https://trello.com/c/v0VZ3WLc/193-bnc-1012968-upgrade-router-migration-took-too-long-during-pre-upgrade-of-one-node)

aspiers reviewed (approved):
LGTM although I would prefer the squash.

rhafer commented:
> LGTM although I would prefer the squash.

Squashed it. And adjusted the commit message.

matelakat commented:
![](https://github.trello.services/images/mini-trello-icon.png) [P2: bsc#1008055: L3: mass migration of neutron routers is very slow](https://trello.com/c/ni7N64ev/1-p2-bsc-1008055-l3-mass-migration-of-neutron-routers-is-very-slow)

matelakat commented:
![](https://github.trello.services/images/mini-trello-icon.png) [P2: bsc#1008055: L3: mass migration of neutron routers is very slow](https://trello.com/c/ni7N64ev/1-p2-bsc-1008055-l3-mass-migration-of-neutron-routers-is-very-slow)
